### PR TITLE
Make IN expr work with multiple items

### LINKED ISSUE #7448: IN expr doesn't work with multiple items under some conditions
### Describe the bug

I noticed `IN` expr doesn't work with multiple items under the following conditions.

1. The left side operand of `IN` is a literal, the number of exprs in the list is greater than 1 and at least one of the exprs includes an attribute reference.
```
CREATE TABLE mytbl(a INT) AS VALUES (1);
SELECT 1 in (a, 2) FROM mytbl;
Execution error: InList expression must evaluate to a scalar
```

2. The left side operand of `IN` is non-literal, at least one of the exprs in the list includes an attribute reference and the number of exprs in the list is greater than 3.
```
CREATE TABLE mytbl(a INT, b INT, c INT, d INT, e INT) AS VALUES (1, 2, 3, 4, 5);
SELECT a IN (b, c, d, e) FROM mytbl;
Execution error: InList expression must evaluate to a scalar
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sarutak**
## Which issue does this PR close?

Closes #7448 

## Rationale for this change

The failure cases shown in #7448 are apparently valid. So this PR is for bug fix.

## What changes are included in this PR?

The cause of the issue is that `in_expr` doesn't consider the case where the evaluated value of each expr in a list varies per row.
So the issue happens in such a case except for the case `IN` expr is simplified by the optimizer.
The fix is to care such a case and search the target value from the list sequentially per the given RecordBatch.

## Are there any user-facing changes?
Yes. But it's a bug fix.

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -326,12 +329,26 @@ impl PhysicalExpr for InListExpr {
     }
 
     fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
-        let value = self.expr.evaluate(batch)?.into_array(1);
+        let value = self.expr.evaluate(batch)?;
         let r = match &self.static_filter {
-            Some(f) => f.contains(value.as_ref(), self.negated)?,
+            Some(f) => f.contains(value.into_array(1).as_ref(), self.negated)?,
```
Comment (sarutak) [inline datafusion/physical-expr/src/expressions/in_list.rs:334]: Oh, I need to fix here for another issue.

**Comment by jonahgao** [comment]
LGTM :+1:.

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -326,12 +329,26 @@ impl PhysicalExpr for InListExpr {
...
             None => {
-                let list = evaluate_list(&self.list, batch)?;
-                make_set(list.as_ref())?.contains(value.as_ref(), self.negated)?
+                let value = value.into_array(batch.num_rows());
+                let found = self.list.iter().map(|expr| expr.evaluate(batch)).try_fold(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:337]: this is a very nice formulation

**Review by alamb — APPROVED** [review]
This looks very nice to me -- well done @sarutak  and thank you for the contribution

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 2 above)
Comment (sarutak) [reply datafusion/physical-expr/src/expressions/in_list.rs:337]: Thanks!

**Pull request merged**